### PR TITLE
Replace example for generally constrained NLOC

### PR DESCRIPTION
**Pull request by hmcm**
A suggestion for replacing the current example for generally constrained NLOC by a more intuitive one: a point mass moving towards a goal in two-dimensional space while avoiding a circular-shaped obstacle.

### CODE REVIEW

File: ct_optcon/examples/NLOC_generalConstrained.cpp [hunk 1]
```diff
@@ -3,180 +3,269 @@
...
 using namespace ct::core;
 using namespace ct::optcon;
 
 
-/*get the state and control input dimension of the oscillator. Since we're dealing with a simple oscillator,
- the state and control dimensions will be state_dim = 2, and control_dim = 1. */
-static const size_t state_dim = ct::core::SecondOrderSystem::STATE_DIM;
-static const size_t control_dim = ct::core::SecondOrderSystem::CONTROL_DIM;
+/* state and control input dimension of the point mass. States: x, y, v_x, v_y.
+ * Controls: linear velocity, angular velocity */
```
Comment (markusgft) [inline ct_optcon/examples/NLOC_generalConstrained.cpp]: Aren't both controls linear velocity?

File: ct_optcon/examples/NLOC_generalConstrained.cpp [hunk 2]
```diff
@@ -3,180 +3,269 @@
...
         Base::ub_.resize(1);
-        Base::lb_.setConstant(-1.0);
-        Base::ub_.setConstant(1.0);
+        Base::lb_.setConstant(0.0);
+        Base::ub_.setConstant(std::numeric_limits<double>::max());
+    }
+
+    ObstacleConstraintSimple(const ObstacleConstraintSimple &other)
+            : Base(other), x0_(other.x0_) {
+        Base::lb_.resize(1);
```
Comment (markusgft) [inline ct_optcon/examples/NLOC_generalConstrained.cpp:92]: this initializer list should not be required, that should be taken care of by the copy-constructor of the base class (but maybe you'd want to confirm that).

File: ct_optcon/examples/NLOC_generalConstrained.cpp [hunk 3]
```diff
@@ -3,180 +3,269 @@
...
+        Base::ub_.resize(1);
+        Base::lb_.setConstant(0.0);
+        Base::ub_.setConstant(std::numeric_limits<double>::max());
+    }
+
+    virtual ~ObstacleConstraintSimple() = default;
```
Comment (markusgft) [inline ct_optcon/examples/NLOC_generalConstrained.cpp]: you can remove the destructor here.

**Review by markusgft — CHANGES_REQUESTED** [review]
Hi @hmcm 
thanks again for that pullrequest, I only had a few small comments, see inline.
One other thing: would you mind quickly running clang-format on your *.cpp and *.h files? It's very straightforward:
```
$ sudo apt install clang-format
$ clang-format -i <path_to_file_that_needs_formatting>
```
It will automatically use the style-file located in the CT root directory.
Thanks!

**Comment by acxz** [comment]
@hmcm why not have both? In my opinion I think it would be great to just add another example instead of replace the existing one.

**Comment by markusgft** [comment]
The old example is unfortunately completely meaningless and unintuitive,
this was the motivation why I agreed with hmcm to replace it...

**Comment by acxz** [comment]
I see, sorry for the noise!

File: ct_optcon/examples/NLOC_generalConstrained.cpp (hunk 1 above)
Comment (hmcm) [reply ct_optcon/examples/NLOC_generalConstrained.cpp]: Actually its forces acting in x- and y-direction... I was still in my differential drive example when I wrote this

File: ct_optcon/examples/NLOC_generalConstrained.cpp (hunk 2 above)
Comment (hmcm) [reply ct_optcon/examples/NLOC_generalConstrained.cpp:92]: Confirmed, thank you!

File: ct_optcon/examples/NLOC_generalConstrained.cpp [hunk 4]
```diff
@@ -74,66 +69,58 @@ class PointMass : public ct::core::ControlledSystem<STATE_DIM, CONTROL_DIM> {
...
-                                     const double t) override {
+    virtual Eigen::VectorXd evaluate(const state_vector_t& x, const control_vector_t& u, const double t) override
+    {
         Eigen::Matrix<double, 1, 1> val;
         // returns a number > 0 if point x is outside of the ellipsoid
-        val.template segment<1>(0) << (1/pow(r_,2)) * (x(0) - x0_(0)) * (x(0) - x0_(0)) +
-                                      (1/pow(r_,2)) * (x(1) - x0_(1)) * (x(1) - x0_(1)) -
-                                      double(1.0);
+        val.template segment<1>(0) << (1 / pow(r_, 2)) * (x(0) - x0_(0)) * (x(0) - x0_(0)) +
+                                          (1 / pow(r_, 2)) * (x(1) - x0_(1)) * (x(1) - x0_(1)) - double(1.0);
```
Comment (markusgft) [inline ct_optcon/examples/NLOC_generalConstrained.cpp]: @hmcm thanks for addressing the comments! I have one more small question: could you quickly verify that the obstacle constraint equation is correct? There is a member variable `r_` that you set to 1.5, but when I look at the final plot, it seems like the curve is much "bigger" than r=1.5. Could you quickly counter-check that the equations here are correct?

Comment (hmcm) [reply ct_optcon/examples/NLOC_generalConstrained.cpp]: @markusgft I double-checked and I think the equation is correct. I agree with you that the curve looks a bit counter-intuitive though. However I am not sure what the cause of this is. For now, I've chosen a time shorter horizon which makes the trajectory more intuitive. I have added the obstacle to the plot to illustrate the results a little better.

Comment (markusgft) [reply ct_optcon/examples/NLOC_generalConstrained.cpp]: Thanks, I will pull them to my hpipm-equipped machine at the next occasion and verify once more, before I merge. Thanks again for the codntribution!

**Comment by markusgft** [comment]
Hi @hmcm, I needed to make a few changes ... I did not find a better way than branching your fork once more and making a new PR.. please find it here #155

**Pull request closed without merging**